### PR TITLE
[1LP][RFR] Automating basic VMRC Console Test

### PR DESCRIPTION
**Pull request by kedark3**
## Purpose or Intent

- __Adding new tests__ VMRC Test automation is added. 

Note: This PR is built on top of work from #9701  and #9595, please review those first so that I can rebase this.

### PRT Run

{{pytest: cfme/tests/infrastructure/test_vmrc_console.py}}

izapolsk: REQUIRES #9764 9764

### CODE REVIEW

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 51db2276a27193e1c4c7393a3d2cba19b9d5a20e

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit b25e835a68f13e8d8859a799c9864381b3e6d336

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit f5a966a7db079562a0aa9eac565cc74b5b0de428

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 9a473094d9c82b996b31366d304d6943bb7b35eb

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 1a2db4f47cf43cd775296c811e91de6934f31591

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 3efd7e4ac565c46d09b6571997dc8b967bf3a936

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit a8422293fcebebdcc30bcd98f91b4272c4cd6e99

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by valaparthvi** [comment]
Restarting the PRT, 510  run resulted in Invalid.

File: cfme/common/vm_console.py [hunk 1]
```diff
@@ -237,7 +237,9 @@ def switch_to_appliance(self):
     def switch_to_console(self):
         """Switch focus to console tab/window."""
         logger.info("Switching to console: window handle = {}".format(self.console_handle))
-        self.browser.selenium.switch_to.window(self.console_handle)
+        if (self.console_handle is not None and self.console_handle in
```
Comment (kedark3) [inline cfme/common/vm_console.py]: In case of VMRC Console, the browser window for console is short-lived, hence it becomes important to check if the `console_handle` exists in available `selenium.window_handles` before attempting to switch to console or close the windw, hence this check.

File: cfme/common/vm_console.py [hunk 2]
```diff
@@ -237,7 +237,8 @@ def switch_to_appliance(self):
     def switch_to_console(self):
         """Switch focus to console tab/window."""
         logger.info("Switching to console: window handle = {}".format(self.console_handle))
-        self.browser.selenium.switch_to.window(self.console_handle)
+        if (self.console_handle and self.console_handle in self.browser.selenium.window_handles):
+            self.browser.selenium.switch_to_window(self.console_handle)
```
Comment (kedark3) [inline cfme/common/vm_console.py]: In case of VMRC Console, the browser window for console is short-lived, hence it becomes important to check if the console_handle exists in available selenium.window_handles before attempting to switch to console or close the windw, hence this check.

File: cfme/tests/infrastructure/test_vmrc_console.py [hunk 3]
```diff
@@ -4,17 +4,53 @@
...
+    vms_collections = appliance.collections.infra_vms
+    vm = vms_collections.instantiate(name='cu-24x7', provider=provider)
+    if not vm.exists_on_provider:
+        pytest.skip("Skipping test, cu-24x7 VM does not exist")
+    if appliance.version < '5.11':
+        vm.open_console(console='VM Console', invokes_alert=True)
+    else:
+        vm.open_console(console='VMRC Console', invokes_alert=True)
+    request.addfinalizer(vm.vm_console.close_console_window)
+    request.addfinalizer(appliance.server.logout)
```
Comment (valaparthvi) [inline cfme/tests/infrastructure/test_vmrc_console.py]: I'm curious, why do you need to explicitly logout here?

File: cfme/common/provider.py [hunk 4]
```diff
@@ -435,7 +435,6 @@ def _fill_smartstate_endpoint_dicts(self, provider_attributes):
 
     def _fill_vmrc_console_endpoint_dicts(self, provider_attributes):
         """Fills dicts with VMRC console endpoint data
-
```
Comment (valaparthvi) [inline cfme/common/provider.py]: Can you remove this change? Kinda seems irrelevant to the PR.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit e54efb823819a41a2cd36939f8be6320792dcc39

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/infrastructure/test_vmrc_console.py (hunk 3 above)
Comment (kedark3) [reply cfme/tests/infrastructure/test_vmrc_console.py]: @valaparthvi without a logout, it was not getting enough time between previous provider removal and next provider addition, neither it would refresh the VMs page to reflect new provider, so logging out seemed best move to avoid unpredictable failures.

File: cfme/common/provider.py (hunk 4 above)
Comment (kedark3) [reply cfme/common/provider.py]: I guess it got introduced during rebase.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 939dd99be66ada26d388f3cb619ae8a8f94b727d

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit fcfff67c4f4c613b86b10539742f86b286b979cc

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/infrastructure/test_vmrc_console.py [hunk 5]
```diff
@@ -2,19 +2,63 @@
...
+    vms_collections = appliance.collections.infra_vms
+    vm = vms_collections.instantiate(name='cu-24x7', provider=provider)
+    if not vm.exists_on_provider:
+        pytest.skip("Skipping test, cu-24x7 VM does not exist")
+    if appliance.version < '5.11':
+        try:
+            vm.open_console(console='VM Console', invokes_alert=True)
+            assert vm.vm_console, 'VMConsole object should be created'
+            request.addfinalizer(vm.vm_console.close_console_window)
+        except ValueError:
```
Comment (jawatts) [inline cfme/tests/infrastructure/test_vmrc_console.py]: If the VM console fails to launch here, we are giving a fast positive. Going to discuss a workaround offline on Monday

Comment (jawatts) [reply cfme/tests/infrastructure/test_vmrc_console.py]: Is the window handle different in 5.10 vs 5.11? Just a thought.

Comment (kedark3) [reply cfme/tests/infrastructure/test_vmrc_console.py]: I think Window Handles are Selenium thing and should have nothing to do with CFME version. Good question though, I will checl.

**Comment by ganeshhubale** [comment]
@kedark3 @valaparthvi I can see in report, tests passed for 5.10 but on PRT. But status of 5.10 is **Invalid**.

File: cfme/common/vm_console.py [hunk 6]
```diff
@@ -252,7 +253,7 @@ def wait_for_connect(self, timeout=30):
 
     def close_console_window(self):
         """Attempt to close Console window at the end of test."""
-        if self.console_handle is not None:
+        if (self.console_handle and self.console_handle in self.browser.selenium.window_handles):
```
Comment (izapolsk) [inline cfme/common/vm_console.py]: `and` has high priority. you don't need braces here

**Comment by dajoRH** [comment]
Lint report for commit 4e43e818ed33d5dc77ac1736b7115398023d48b0:

`cfme/tests/infrastructure/test_vmrc_console.py`:
- :red_circle: Line 5:1: **F401** *'cfme.utils.log.logger' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 4e43e818ed33d5dc77ac1736b7115398023d48b0

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 6510614a1f8ab3da57af16e14016a7f5cec8ea5a

<details>
<summary>
<b>Show fixtures</b>
</summary>


The *global* fixture **`generic_catalog_item_with_imported_dialog`** is used in the following files:
- **cfme/tests/services/test_dialog_element_in_catalog.py**
    - *test_dialog_dropdown_int_required*

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

The *local* fixture **`config_system`** is used in the following files:
- **cfme/tests/infrastructure/test_config_management.py**
    - *test_config_system_tag*
    - *test_config_system_reprovision*

The *local* fixture **`authentications`** is used in the following files:
- **cfme/tests/infrastructure/test_config_management_rest.py**
    - *_check_edited_authentications*
    - *test_query_authentications_attributes*
    - *test_authentications_edit_single*
    - *test_authentications_edit_multiple*
    - *test_delete_authentications_from_detail_post*
    - *test_delete_authentications_from_detail_delete*
    - *test_delete_authentications_from_collection*

The *local* fixture **`config_manager_rest`** is used in the following files:
- **cfme/tests/infrastructure/test_config_management_rest.py**
    - *test_config_manager_create_rest*
    - *test_config_manager_edit_rest*
    - *test_config_manager_delete_rest*

The *local* fixture **`ansible_workflow_catitem`** is used in the following files:
- **cfme/tests/services/test_ansible_workflow_servicecatalogs.py**
    - *test_tower_workflow_item*
    - *test_retire_ansible_workflow*

The *local* fixture **`catalog_item`** is used in the following files:
- **cfme/tests/services/test_config_provider_servicecatalogs.py**
    - *test_order_tower_catalog_item*
    - *test_retire_ansible_service*

The *local* fixture **`change_company_name`** is used in the following files:
- **cfme/tests/test_providerless_rest.py**
    - *test_update_roles_via_rest_name_change*

</details>

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f3f0503680f4387c41591dfaa6fb4ea5d5a3263e:

`cfme/tests/infrastructure/test_vmrc_console.py`:
- :red_circle: Line 5:1: **F401** *'cfme.utils.log.logger' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit ad4ed14991bc195db830ef5bf5f2c625b725765a

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by kedark3** [comment]
@mshriver @john-dupuy Tests are now passing in PRT thanks to @izapolsk . Kindly review and let me know if we can merge it. 510z passed but PRT seems to be stuck, cfme.log indicate test was passed.

File: cfme/common/vm.py [hunk 7]
```diff
@@ -275,13 +275,19 @@ def open_console(self, console='VM Console', invokes_alert=None):
                            setting this to true will handle this.
         """
         # TODO: implement vmrc vm console
```
Comment (john-dupuy) [inline cfme/common/vm.py]: This `TODO` can now be removed

File: cfme/common/vm.py [hunk 8]
```diff
@@ -275,13 +275,19 @@ def open_console(self, console='VM Console', invokes_alert=None):
...
         view = navigate_to(self, 'Details')
 
-        # Click console button given by type
-        view.toolbar.access.item_select(console, handle_alert=invokes_alert)
+        # dismiss_any_alerts() call closed subsequent alerts needed for vmrc
+        # below code is needed to fix such issue
+        try:
+            view.browser.IGNORE_SUBSEQUENT_ALERTS = True
+            # Click console button given by type
+            view.toolbar.access.item_select(console, handle_alert=invokes_alert)
```
Comment (john-dupuy) [inline cfme/common/vm.py]: Why the name change from `handle_alert` to `invokes_alert`? I would keep it consistent as `handle_alert`

File: cfme/tests/infrastructure/test_vmrc_console.py [hunk 9]
```diff
@@ -4,17 +4,60 @@
...
+        pytest.skip("Skipping test, cu-24x7 VM does not exist")
+    if appliance.version < '5.11':
+        # try:
+        vm.open_console(console='VM Console', invokes_alert=True)
+        assert vm.vm_console, 'VMConsole object should be created'
+        request.addfinalizer(vm.vm_console.close_console_window)
+        # except ValueError:
+        #     # with 5.10, console window closes too quickly to catch vm_console handle
+        #     # hence we just check if we could open_console, not validating VMconsole object
+        #     logger.info("Could not get the vm.vm_console, window closed too quickly, ignoring.")
```
Comment (john-dupuy) [inline cfme/tests/infrastructure/test_vmrc_console.py]: Can this be removed?

File: cfme/tests/infrastructure/test_vmrc_console.py [hunk 10]
```diff
@@ -4,17 +4,60 @@
...
+            2. You should see a Pop-up being Blocked, Please allow it to
+               open (always allow pop-ups for this site) and then a new tab
+               will open and then in few seconds for the VMRC console.
+    """
+    vms_collections = appliance.collections.infra_vms
+    vm = vms_collections.instantiate(name='cu-24x7', provider=provider)
+    if not vm.exists_on_provider:
+        pytest.skip("Skipping test, cu-24x7 VM does not exist")
+    if appliance.version < '5.11':
+        # try:
```
Comment (john-dupuy) [inline cfme/tests/infrastructure/test_vmrc_console.py]: Can this be removed?

File: cfme/common/vm.py (hunk 8 above)
Comment (kedark3) [reply cfme/common/vm.py]: This is not being changed in the PR, its been like that always. And to me it makes sense because, when I am calling it like:
`vm.open_console(console='VM Console', invokes_alert=True)` 
it reads like, open VM console and that console invokes_alert. I think we can keep it as is unless you insist.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit bb734c9badd1fc9a8ba199c02939109858d3cb2d

The *global* fixture **`configure_console_vmrc`** is used in the following files:
- **cfme/tests/infrastructure/test_vmrc_console.py**
    - *test_vmrc_console*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/common/vm.py (hunk 8 above)
Comment (john-dupuy) [reply cfme/common/vm.py]: That's fine with me, if you like it then we can keep it as is.
